File: upwind/python/saveVTK.py
#!/usr/bin/env python

def rectilinear(fname, xAxis, yAxis, zAxis, data):
  """
  Save data in reclilinear VTK file format
  @param fname file name
  @param xAxis x axis
  @param yAxis y axis
  @param zAxis z axis
  @param data 
  """

  numCells = data.shape
  ndims = len(numCells)

  with open(fname, 'w') as f:
    f.write("# vtk DataFile Version 2.0\n")
    f.write("upwind\n")
    f.write("ASCII\n")
    f.write("DATASET RECTILINEAR_GRID\n")
    f.write("DIMENSIONS")
    # in VTK the first dimension varies fastest so need 
    # to invert the order of the dimensions
    if ndims > 2:
      f.write(f' {numCells[2] + 1}')
    else:
      f.write(" 1")
    if ndims > 1:
      f.write(f' {numCells[1] + 1}')
    else:
      f.write(" 1")
    f.write(f' {numCells[0] + 1}')
    f.write("\nX_COORDINATES ")
    if ndims > 2:
      f.write(f"{numCells[2] + 1} double")
      for i in range(numCells[2] + 1):
        f.write(f' {xAxis[i]}')
    else:
      f.write("1 double 0.0")
    f.write( "\nY_COORDINATES ")
    if ndims > 1:
      f.write(f"{numCells[1] + 1} double")
      for i in range(numCells[1] + 1): 
        f.write(f' {yAxis[i]}') 
    else:
      f.write("1 double 0.0")
    f.write("\nZ_COORDINATES ")
    f.write(f"{numCells[0] + 1} double")
    for i in range(numCells[0] + 1):
      f.write(f' {zAxis[i]}')

    flatData = data.flat
    ntot = len(flatData)

    f.write(f"\nCELL_DATA {ntot}\n")
    f.write("SCALARS f double 1\n")
    f.write("LOOKUP_TABLE default\n")
      
    for i in range(len(flatData)):
      f.write(f'{flatData[i]}')
      if (i + 1) % 10 == 0:
          f.write('\n')

File: upwind/julia/saveVTK.jl
module saveVTK

export rectilinear

function rectilinear(fname::AbstractString,
                     xAxis::LinSpace, yAxis::LinSpace, zAxis::LinSpace,
                     data::Array)
    """
    Save data on rectilinear grid in VTK file
    @param fname file name
    @param xAxis x axis (nodes)
    @param yAxis y axis (nodes)
    @param zAxis z axis (nodes)
    @param data field data
    """
    f = open(fname, "w")
    write(f, "# vtk DataFile Version 2.0\n")
    write(f, "upwind.jl\n")
    write(f, "ASCII\n")
    write(f, "DATASET RECTILINEAR_GRID\n")
    write(f, "DIMENSIONS")

    numCells = [length(xAxis) - 1, length(yAxis) - 1, length(zAxis) - 1]
    ntot = length(data)

    # Number of nodes in each direction
    write(f, @sprintf(" %d", numCells[1] + 1))
    write(f, @sprintf(" %d", numCells[2] + 1))
    write(f, @sprintf(" %d\n", numCells[3] + 1))

    write(f,  "X_COORDINATES ")
    write(f,  @sprintf("%d double\n", numCells[1] + 1))
    for i in 1:numCells[1] + 1
        write(f,  @sprintf("%f ", xAxis[i]))
    end 

    write(f,  "\nY_COORDINATES ")
    write(f, @sprintf("%d double\n", numCells[2] + 1))
    for i in 1:numCells[2] + 1 
        write(f, @sprintf("%f ", yAxis[i]))
    end

    write(f,  "\nZ_COORDINATES ")
    write(f, @sprintf("%d double\n", numCells[3] + 1))
    for i in 1:numCells[3] + 1
        write(f, @sprintf("%f ", zAxis[i]))
    end

    write(f, @sprintf("\nCELL_DATA %d\n", ntot))
    write(f, "SCALARS f double 1\n")
    write(f, "LOOKUP_TABLE default\n")
    for i in eachindex(data)
        write(f, @sprintf("%f ", data[i]))
        if i % 10 == 0
            write(f, "\n")
        end
    end
    write(f, "\n")

    close(f)


end

end
